src/parse.jl

```diff
@@ -0,0 +1,254 @@
+"""
+    count_locations_and_transitions(root_sxmodel)
+
+Returns the number of locations and transitions for each component.
+
+### Input
+
+- `root_sxmodel` -- the root element of a SX file
+
+### Output
+
+Two vectors of integers `(lcount, tcount)`, where the i-th entry of `lcount` and
+`tcount` are number of locations and transitions, respectively, of the i-th component.
+"""
+function count_locations_and_transitions(root_sxmodel)
+    lcount = Vector{Int}() # num locations for each component
+    tcount = Vector{Int}() # num transitions for each component
+    id_component = 0 # index of the current component
+
+    # count the number of locations for each component
+    for component in eachelement(root_sxmodel)
+        if nodename(component) == "component"
+            id_component += 1
+            push!(lcount, 0)
+            push!(tcount, 0)
+            for field in eachelement(component)
+                if nodename(field) == "location"
+                    lcount[id_component] += 1
+                elseif nodename(field) == "transition"
+                    tcount[id_component] += 1
+                end
+            end
+        end
+    end
+    return (lcount, tcount)
+end
+
+"""
+    parse_sxmath(s)
+
+Returns the list of expressions corresponding to a given SX string.
+
+### Input
+
+- `s` - string
+
+### Output
+
+Vector of expressions, equations or inequalities.
+
+### Examples
+
+```jldoctest
+julia> import SX.parse_sxmath
+
+julia> parse_sxmath("x >= 0")
+1-element Array{Expr,1}:
+ :(x >= 0)
+
+julia> parse_sxmath("x' == x & v' == -0.75*v")
+2-element Array{Expr,1}:
+ :(x' = x)
+ :(v' = -0.75v)
+
+julia> parse_sxmath("x == 0 & v <= 0")
+2-element Array{Expr,1}:
+ :(x = 0)
+ :(v <= 0)
+```
+
+### Algorithm
+
+Consists of the following steps (in the given order):
+
+- split the string with the `&` key
+- remove trailing whitespace of each substring
+- replace double `==` with single `=`
+- cast to a Julia expression with `parse`
+"""
+parse_sxmath(s) = parse.(replace.(strip.(split(s, "&")), "==", "="))
+
+"""
+    parse_sxmodel!(root_sxmodel, HDict)
+
+### Input
+
+- `HDict` - dictionary that wraps the hybrid model and contains the keys `(automaton,
+            variables, transitionlabels, invariants, flows, resetmaps, switchings)`
+
+### Output
+
+The `HDict` dictionary.
+
+### Notes
+
+Edge labels are not used and their symbol is (arbitrarily) set to the integer 1.
+"""
+function parse_sxmodel!(root_sxmodel, HDict)
+
+    # edge labels
+    σ = 1
+
+    # counter for the variables
+    id_variable = 0
+
+    # counter for the transitions
+    id_transition = 0
+
+    for component in eachelement(root_sxmodel)
+        for field in eachelement(component)
+            if nodename(field) == "param"
+                if field["type"] == "real"
+                    id_variable += 1
+                    add_variable!(HDict["variables"], field, id_variable)
```

Comment:
Does it make sense to add fresh variables for each component? Usually the variables should be shared.
Maybe add a comment for future extensions here.

Reply:
this is a good point. @nikos-kekatos correct me if i'm wrong but i think that in general (> 1 components), variables in each component are different by default, like components are different scopes in a program, but then you define "bindings" to match variables in different components.

let me add a comment for future extensions.